### DISCUSSION TITLE
Use TinySDF to generate missing glyphs in the client

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre-gl-js#2555] [Ideas]
Sometimes a font file does not contain a glyph. Say for example that you have a font which covers `a-z` and `A-Z` but nothing else. Then if you want to render `Zürich`, the default behavior of MapLibre GL JS is to show `Zrich` I think. So it just skips the missing `ü`. In such cases where a glyph is missing from the font, we could generate it dynamically in the client with TinySDF.

**Comment by wipfli** [comment]
What do you think? cc @1ec5 and @ZeLonewolf because I think you spent quite some time putting a font together which has maximal coverage for Americana...

**Reply by ZeLonewolf** [reply]
I don't think I'm smart enough on font issues yet to offer an opinion here.

**Comment by 1ec5** [comment]
Your assumption is correct: osm-americana/openstreetmap-americana#850 has some examples of what happens when the glyph PBFs lack required glyphs.

From an end-user perspective, your proposal is good for basic usability, assuming that the `text-field` value isn’t garbage. From a developer perspective, I don’t see a backwards compatibility issue at a glance, but it deserves a deeper look. Font selection is already so complex that preserving compatibility with Mapbox software might be somewhat important.

The traditional fontstack model always seemed unreasonable to me, especially that the fontstack would get flattened on the server side into a single set of PBFs, requiring the style to specify a pan-Unicode, last-resort fallback font as part of the fontstack. A last-last-resort fallback on the client side would reduce the dependency on fontstacks, getting us closer to the point where we can ditch that concept and, along with it, the faulty assumptions in https://github.com/maplibre/maplibre-native/discussions/778#discussioncomment-5051041.

As a side benefit, if developers who roll their own glyph PBFs can assume a version of MapLibre that has the proposed behavior, then they can roll much smaller PBFs that contain only the characters from specialty fonts and none of the fallback characters from boring pan-Unicode fonts. But a bit of caution is warranted: whenever a combining character’s font mismatches the surrounding text’s font, text layout always gets wonky.

**Reply by 1ec5** [reply]
> A last-last-resort fallback on the client side would reduce the dependency on fontstacks, getting us closer to the point where we can ditch that concept and, along with it, the faulty assumptions in https://github.com/maplibre/maplibre-native/discussions/778#discussioncomment-5051041.

Related: maplibre/maplibre-gl-js#4564, maplibre/maplibre-gl-js#4550, 1ec5/maplibre-gl-js#1, osm-americana/openstreetmap-americana#1149.

**Comment by spatialillusions** [comment]
I have a special use case where I run maplibre offline as a single HTML page reading a pmtiles file, for me it would be great if I just could use local fonts instead of having to make sure to copy all glyph files to the clients. At the moment I have made sure to include some char sets into the source and a custom protocol for reading them, but there are always some some glyphs missing.
Can't wait until this is merged into maplibre.

**Reply by HarelM** [reply]
I have a similar use case, but using local glyphs folders with the relevant fonts is behaving as expected, for Latin and Hebrew only though.

**Reply by spatialillusions** [reply]
I solved it! 
Let me just clean up my dirty JavaScript and I will share it, but the base concept is a custom protocol for glyphs, that calls my code that uses pbf and tiny-sdf to create glyph-ranges in the browser and hand them back to maplibre as if they were pbf files loaded from a server or disk.

**Reply by spatialillusions** [reply]
Dumped my code here:
https://github.com/spatialillusions/maplibre-local-glyphs

**Reply by spatialillusions** [reply]
Now available as a NPM package. 
https://www.npmjs.com/package/maplibre-local-glyphs
Please let me know if it works for you.

**Reply by HarelM** [reply]
Interesting, I'll see if I can use that to replace the sprite as well.

**Reply by spatialillusions** [reply]
I have another solution for spirits where I store them as base64 encoded strings. Ping me on the email on my profile if you would like the source code. (It's a bit rough.)

**Reply by HarelM** [reply]
No, I'm storing the icon of my site in a font, like what's done in font-awesome.
So this might be a way to share icons between the map and the website.

**Reply by HarelM** [reply]
Also it's probably worth adding this plugin to awesome-maplibre repo so that it will appear in the docs and elsewhere.

**Reply by HarelM** [reply]
A PR for local glyphs was recently merged and there's a plan to support `font-faces` in the future from a style spec perspective (already supported in native) so it might be that this plugin won't be necessary in the near future, which I think is good :-)
